lib/resty/etcd/cluster/healthcheck.lua

```diff
@@ -0,0 +1,202 @@
+            http_statuses = {200, 201},
+            successes = 1,
+        },
+        unhealthy = {
+            http_statuses = {500},
+            http_failures = 1,
+            tcp_failures = 1,
+            --passive health checks are disabled by default
+            timeouts = 0,
+        },
+    },
+}, fixed_field_metatable)
+
+
+local _M = {
+    version = 0.1,
+}
+
+
+function _M.report_failure(endpoint, osi, _, status)
+    if not checker then
+        return
+    end
+
+    if osi == "tcp" then
+        checker:report_tcp_failure(endpoint.host, tonumber(endpoint.port),
+                nil, nil, "active")
+        return
+    end
+
+    if osi == "http" then
+        if status >= 500 then
+            checker:report_http_status(endpoint.host, tonumber(endpoint.port),
+                    nil, status, "active")
+            return
+        end
+    end
+
+end
+
+
+function _M.fetch_health_nodes(endpoints)
+    if not checker then
+        return nil
+    end
+
+    for _, endpoint in ipairs(endpoints) do
+        local ok = checker:get_target_status(endpoint.host, tonumber(endpoint.port))
+        if ok then
+            return endpoint
+        end
+    end
+
+    utils.log_error("etcd cluster is unavailable")
+    return nil
+end
+
+
+function _M.run(opts, endpoints)
+    local shared_dict = ngx_shared[opts.cluster_healthcheck.shm_name]
+    if not shared_dict then
+        utils.log_error("failed to get ngx.shared dict when start etcd cluster health check")
+        return
+    end
+
+    --supported etcd version >= 3.3.0
+    --https://github.com/etcd-io/etcd/blob/master/Documentation/op-guide/monitoring.md#health-check
+    if not headthcheck_endpoint[opts.api_prefix] then
+        utils.log_info("unsupported health check for the etcd version < v3.3.0")
+        return
+    end
+
+    if not checker then
+        local ok, checks = pcall(tbl_copy_merge_defaults, opts.cluster_healthcheck.checks, DEFAULTS)
+        if not ok then
+            utils.log_error("err: ", checks)
+            return nil, checks
+        end
+
+        if healthcheck == nil then
+            healthcheck = require("resty.healthcheck")
+        end
+
+        checker = healthcheck.new({
+            name = opts.cluster_healthcheck.name or "etcd-cluster-health-check",
+            shm_name = opts.cluster_healthcheck.shm_name,
+            checks = checks,
+        })
+
+        if #endpoints > 1 then
+            for _, endpoint in ipairs(endpoints) do
+                local _, err = checker:add_target(endpoint.host, endpoint.port, nil, true)
+                if not ok then
+                    utils.log_error("failed to add new health check target: ", endpoint.host, ":",
+                            endpoint.port, " err: ", err)
+                end
+                utils.log_info("success to add new health check target: ", endpoint.host, ":",
+                        endpoint.port)
+            end
+        end
+
+        checker:start()
```

Comment:
Should be `return checker:start()`

Reply:
if `return checker:start()`, then `local checker = healthcheck.run(opts, endpoints)` , the `checker` will be `nil`.
I thought use `return checker:start()`  was used to pass out the `checker` pointer.

Reply:
We need a way to know if the `check:start()` successes

Reply:
i know

Reply:
Should be `local ok, err  = checker:start()`, and return the `nil, err` if needed

Reply:
I changed this way, pls review.